SofaKernel/modules/SofaExplicitOdeSolver/EulerSolver.cpp

```diff
@@ -46,92 +46,132 @@ using namespace sofa::defaulttype;
+    , d_optimizedForDiagonalMatrix(initData(&d_optimizedForDiagonalMatrix, true, "optimizedForDiagonalMatrix", "If true, solution to the system Ax=b can be directly found by computing x = f/m. Must be set to false if M is sparse."))
     , d_threadSafeVisitor(initData(&d_threadSafeVisitor, false, "threadSafeVisitor", "If true, do not use realloc and free visitors in fwdInteractionForceField."))
 {
 }
 
 typedef simulation::Visitor::ctime_t ctime_t;
 
-void EulerSolver::solve(const core::ExecParams* params, SReal dt, sofa::core::MultiVecCoordId xResult, sofa::core::MultiVecDerivId vResult)
+void EulerExplicitSolver::solve(const core::ExecParams* params, SReal dt, sofa::core::MultiVecCoordId xResult, sofa::core::MultiVecDerivId vResult)
 {
     sofa::simulation::common::VectorOperations vop( params, this->getContext() );
     sofa::simulation::common::MechanicalOperations mop( params, this->getContext() );
     mop->setImplicit(false); // this solver is explicit only
     MultiVecCoord pos(&vop, core::VecCoordId::position() );
     MultiVecDeriv vel(&vop, core::VecDerivId::velocity() );
-    MultiVecDeriv acc(&vop, core::VecDerivId::dx()); acc.realloc(&vop, !d_threadSafeVisitor.getValue(), true); // dx is no longer allocated by default (but it will be deleted automatically by the mechanical objects)
     MultiVecDeriv f  (&vop, core::VecDerivId::force() );
-    MultiVecCoord pos2(&vop, xResult /*core::VecCoordId::position()*/ );
-    MultiVecDeriv vel2(&vop, vResult /*core::VecDerivId::velocity()*/ );
 
-    mop.addSeparateGravity(dt); // v += dt*g . Used if mass wants to added G separately from the other forces to v.
-    sofa::helper::AdvancedTimer::stepBegin("ComputeForce");
-    mop.computeForce(f);
-    sofa::helper::AdvancedTimer::stepEnd("ComputeForce");
+    MultiVecCoord newPos(&vop, xResult /*core::VecCoordId::position()*/ );
+    MultiVecDeriv newVel(&vop, vResult /*core::VecDerivId::velocity()*/ );
+    MultiVecDeriv acc(&vop, core::VecDerivId::dx());
 
-    sofa::helper::AdvancedTimer::stepBegin("AccFromF");
-    mop.accFromF(acc, f);
-    sofa::helper::AdvancedTimer::stepEnd("AccFromF");
-    mop.projectResponse(acc);
+    acc.realloc(&vop, !d_threadSafeVisitor.getValue(), true); // dx is no longer allocated by default (but it will be deleted automatically by the mechanical objects)
+
+    // Mass matrix is diagonal, solution can thus be found by computing acc = f/m
+    if(d_optimizedForDiagonalMatrix.getValue())
+    {
+        mop.addSeparateGravity(dt); // v += dt*g . Used if mass wants to add G separately from the other forces to v.
+        sofa::helper::AdvancedTimer::stepBegin("ComputeForce");
+        mop.computeForce(f);
+        sofa::helper::AdvancedTimer::stepEnd("ComputeForce");
+
+        msg_info() << "f = "<< f;
```

Comment:
Could we output the norm of f instead of the full vector? It is simply awful to debug when you have huge vector displayed in the output stream...